File: dbt-runner/app/core/host_guard.py
"""Allow/deny checks for connection targets chosen by a user.

Every connection test and every ingest job points dbt-runner at a host the
requesting user typed in. Without a check, a user can aim one at this
deployment's own Postgres, read `connections.password_encrypted` for every
other user, and walk away with their warehouse credentials - a privilege
escalation through an entirely legitimate-looking UI flow. Cloud metadata
endpoints (169.254.169.254) are the same story for instance credentials.

Used by both the connection router and the ingest runner: a guard on only one
of the two paths is a guard nobody can rely on.
"""

import ipaddress
import logging
import socket
from typing import Iterable
from urllib.parse import urlparse

from app.config import settings

logger = logging.getLogger(__name__)


class HostNotAllowed(ValueError):
    """Raised when a target host is refused by policy."""


# Cloud instance-metadata services. Reachable from inside a container and hand
# out credentials to anyone who asks, so these are never allowed.
_METADATA_NETWORKS = (
    ipaddress.ip_network("169.254.0.0/16"),
    ipaddress.ip_network("fe80::/10"),
)

_LOOPBACK_HINT = (
    "loopback resolves to dbt-runner itself inside the container, not to your "
    "database host - use the service name or LAN address"
)


def _endpoint(url: str) -> tuple[str, int | None] | None:
    """Extract (host, port) from a connection URL, ignoring unparseable input."""
    if not url:
        return None
    try:
        parsed = urlparse(url)
    except ValueError:
        return None
    if not parsed.hostname:
        return None
    try:
        port = parsed.port
    except ValueError:
        port = None
    return parsed.hostname.lower(), port


def own_infrastructure() -> set[tuple[str, int | None]]:
    """Endpoints belonging to this deployment, which no source may target."""
    endpoints = set()
    for url in (settings.database_url, settings.lake_catalog_url, settings.redis_url):
        endpoint = _endpoint(url)
        if endpoint:
            endpoints.add(endpoint)
    return endpoints


def _resolve(host: str) -> list[ipaddress.IPv4Address | ipaddress.IPv6Address]:
    """Resolve a hostname to every address it currently answers with."""
    try:
        return [ipaddress.ip_address(host)]
    except ValueError:
        pass

    try:
        infos = socket.getaddrinfo(host, None)
    except socket.gaierror as exc:
        raise HostNotAllowed(f"Host '{host}' could not be resolved: {exc}") from exc

    addresses = []
    for info in infos:
        try:
            addresses.append(ipaddress.ip_address(info[4][0]))
        except ValueError:
            continue
    if not addresses:
        raise HostNotAllowed(f"Host '{host}' resolved to no usable address")
    return addresses


def _reject(host: str, reason: str) -> None:
    # The message reaches the user, so it names the host they typed and the
    # policy that refused it - never the internal endpoint it collided with.
    raise HostNotAllowed(f"Target host '{host}' is not allowed: {reason}")


def assert_host_allowed(
    host: str,
    port: int | None = None,
    *,
    own: Iterable[tuple[str, int | None]] | None = None,
) -> None:
    """Refuse hosts that reach this deployment's own infrastructure.

    Checks the *resolved* addresses rather than the hostname, so a name that
    points at 127.0.0.1 or the metadata service is caught regardless of what it
    is called.

    ponytail: resolves for the check, but the client re-resolves when it
    connects, so a racing DNS change can still slip past. Pinning the address
    into each connection string is the fix if this ever guards untrusted users
    rather than employees of one company.
    """
    host = (host or "").strip().lower()
    if not host:
        _reject(host, "no host given")

    addresses = _resolve(host)

    # Compare resolved addresses, not names: with INGEST_ALLOW_PRIVATE_HOSTS on
    # - which every on-premise deployment needs - a name check alone is bypassed
    # by typing the app database's LAN IP instead of its service name.
    for infra_host, infra_port in own if own is not None else own_infrastructure():
        if port is not None and infra_port is not None and port != infra_port:
            continue
        if host == infra_host:
            _reject(host, "it is this deployment's own database or cache")
        try:
            infra_addresses = set(_resolve(infra_host))
        except HostNotAllowed:
            continue  # own endpoint unresolvable here; the name check above still applies
        if infra_addresses & set(addresses):
            _reject(host, "it is this deployment's own database or cache")

    for address in addresses:
        if any(address in network for network in _METADATA_NETWORKS):
            _reject(host, "link-local and cloud metadata addresses are blocked")
        if address.is_loopback:
            _reject(host, _LOOPBACK_HINT)
        if address.is_private and not settings.ingest_allow_private_hosts:
            _reject(
                host,
                "private network addresses are blocked - set "
                "INGEST_ALLOW_PRIVATE_HOSTS=true for on-premise warehouses",
            )

File: nextjs/src/lib/host-guard.ts
import { lookup } from "node:dns/promises"
import { isIP } from "node:net"

/**
 * Refuse URLs that point back into this deployment's own network.
 *
 * A model provider's Base URL is typed by the user and then fetched by the
 * server, which makes it the same surface dbt-runner's `app/core/host_guard.py`
 * guards for connection hosts: without a check, `http://169.254.169.254/...`
 * returns instance credentials and `http://postgres:5432` probes the app
 * database, with the response handed back to the browser.
 *
 * Addresses are checked *after* resolution, so a public name that answers with
 * 127.0.0.1 is caught regardless of what it is called.
 *
 * ponytail: resolves for the check, but fetch() resolves again when it
 * connects, so a racing DNS change can still slip past. Pinning the resolved
 * address into the request is the fix if this ever guards untrusted users
 * rather than colleagues.
 */

export class HostNotAllowed extends Error {}

/**
 * On-premise deployments run their gateway on the LAN, so private ranges are a
 * setting rather than a rule. Loopback and link-local never open: loopback is
 * the frontend container itself, and link-local is the metadata service.
 */
function privateHostsAllowed(): boolean {
  return process.env.AI_PROVIDER_ALLOW_PRIVATE_HOSTS === "true"
}

function ipv4Rules(address: string): string | null {
  const parts = address.split(".").map(Number)
  if (parts.length !== 4 || parts.some((part) => !Number.isInteger(part) || part < 0 || part > 255)) {
    return "it is not a usable address"
  }
  const [a, b] = parts
  if (a === 127) return "loopback is this container itself, not your gateway host"
  if (a === 0) return "it is not a usable address"
  if (a === 169 && b === 254) return "link-local and cloud metadata addresses are blocked"
  if (a === 10) return "it is a private address"
  if (a === 172 && b >= 16 && b <= 31) return "it is a private address"
  if (a === 192 && b === 168) return "it is a private address"
  if (a === 100 && b >= 64 && b <= 127) return "it is a carrier-private address"
  return null
}

function ipv6Rules(address: string): string | null {
  const value = address.toLowerCase().split("%")[0]
  const mapped = value.match(/^::ffff:(\d+\.\d+\.\d+\.\d+)$/)
  if (mapped) return ipv4Rules(mapped[1])
  // WHATWG URL re-serialises an IPv4-mapped literal into hex groups, so
  // ::ffff:169.254.169.254 arrives here as ::ffff:a9fe:a9fe.
  const mappedHex = value.match(/^::ffff:([0-9a-f]{1,4}):([0-9a-f]{1,4})$/)
  if (mappedHex) {
    const high = parseInt(mappedHex[1], 16)
    const low = parseInt(mappedHex[2], 16)
    return ipv4Rules([high >> 8, high & 255, low >> 8, low & 255].join("."))
  }
  if (value === "::1") return "loopback is this container itself, not your gateway host"
  if (value === "::") return "it is not a usable address"
  if (/^fe[89ab]/.test(value)) return "link-local and cloud metadata addresses are blocked"
  if (/^f[cd]/.test(value)) return "it is a private address"
  return null
}

/** Every reason to refuse an address, or null when it is reachable and public. */
function refusalFor(address: string): string | null {
  return isIP(address) === 6 ? ipv6Rules(address) : ipv4Rules(address)
}

async function resolveAll(host: string): Promise<string[]> {
  if (isIP(host)) return [host]
  try {
    const records = await lookup(host, { all: true })
    if (records.length === 0) throw new Error("no address")
    return records.map((record) => record.address)
  } catch {
    throw new HostNotAllowed(`Host “${host}” could not be resolved from this deployment.`)
  }
}

export async function assertUrlHostAllowed(url: string): Promise<void> {
  let parsed: URL
  try {
    parsed = new URL(url)
  } catch {
    throw new HostNotAllowed("Base URL must be a valid http(s) URL")
  }
  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    throw new HostNotAllowed("Base URL must be a valid http(s) URL")
  }
  // new URL keeps the brackets of an IPv6 literal; the address is what resolves.
  const host = parsed.hostname.replace(/^\[|\]$/g, "")

  for (const address of await resolveAll(host)) {
    const refusal = refusalFor(address)
    if (!refusal) continue
    const isPrivate = refusal.includes("private")
    if (isPrivate && privateHostsAllowed()) continue
    throw new HostNotAllowed(
      isPrivate
        ? `Base URL “${host}” is not allowed: ${refusal}. Set AI_PROVIDER_ALLOW_PRIVATE_HOSTS=true to reach a gateway on this network.`
        : `Base URL “${host}” is not allowed: ${refusal}.`,
    )
  }
}
